### PR TITLE
[chenguo3125] iP

### PR DESCRIPTION
**Pull request by chenguo3125**
# 🐰 Usagi: A Smart Task Management Chatbot

## Overview
**Usagi** is an intelligent command-line and GUI task management application that helps users organize their daily tasks, deadlines, and events through natural language commands.

## ✨ Key Features

- **Task Types**: Support for `todo`, `deadline`, and `event` tasks
- **Smart Date Parsing**: Flexible date/time input formats with `java.time` API
- **Persistent Storage**: Automatic saving and loading of tasks from file
- **Search Functionality**: Find tasks by keyword using the `find` command
- **Date Filtering**: View tasks occurring on specific dates with `on <date>` command
- **Dual Interface**: Both command-line and JavaFX GUI support

## 🏗️ Architecture

The project follows clean architecture principles with clear separation of concerns:

1. **`usagi.ui`** - User interface and interaction handling
2. **`usagi.storage`** - File I/O and data persistence
3. **`usagi.parser`** - Command interpretation and validation
4. **`usagi.task`** - Task models and business logic
5. **`usagi.exception`** - Custom exception handling

## 💻 Usage Examples

```java
// Create a deadline task
deadline return book /by 2024-12-25 1800

// Find tasks containing "book"
find book

// View tasks on a specific date
on 2024-12-25

// Mark task as complete
mark 1
```

## 🧪 Testing

- [x] **Unit Tests**: Comprehensive JUnit 5 test coverage
- [x] **Integration Tests**: End-to-end functionality validation
- [x] **Edge Cases**: Error handling and boundary condition testing
- [ ] **Performance Tests**: Load testing for large task lists

### Prerequisites
- Java 11 or higher
- Gradle 8.5+

### Build and Run
```bash
# Build the project
./gradlew build

# Create executable JAR
./gradlew shadowJar

# Run the application
java -jar build/libs/usagi.jar
```

## 🔧 Technical Implementation

The application leverages modern Java features including:
- **`java.time` API** for robust date/time handling
- **Gradle** for build automation and dependency management
- **Package organization** for maintainable code structure
- **Exception handling** for graceful error management
- **JavaFX** for modern GUI interface

## 📝 Code Quality

> *"Clean code is simple and direct. Clean code reads like well-written prose."* - Robert C. Martin

- **JavaDoc**: Comprehensive documentation for all public APIs
- **Coding Standards**: Consistent with Java coding conventions
- **Encapsulation**: Protected fields with controlled access
- **Modularity**: Clear separation of concerns across classes

## 🔗 Related Resources

- [Java Time API Documentation](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/time/package-summary.html)
- [Gradle User Guide](https://docs.gradle.org/current/userguide/userguide.html)
- [JUnit 5 User Guide](https://junit.org/junit5/docs/current/user-guide/)

## 🎯 Future Enhancements

- [ ] **Priority Levels**: Add task priority management
- [ ] **Categories/Tags**: Organize tasks by categories
- [ ] **Recurring Tasks**: Support for repeating tasks
- [ ] **Export/Import**: Data portability features
- [ ] **Enhanced GUI**: More interactive JavaFX components

---

*Built with ❤️ using Java and modern software engineering practices*

### CODE REVIEW

File: src/main/java/usagi/parser/Parser.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        } else if (input.startsWith("find ")) {
+            String keyword = input.substring(5).trim();
+            if (keyword.isEmpty()) {
+                ui.showLine("Ura? (find must be followed by a keyword)");
+            } else {
+                List<Task> matches = tasks.find(keyword);
+                ui.showFound(matches);
+            }
+            return;
+        } else if (input.contains("mark")) {
```
Comment (jooh6969) [inline src/main/java/usagi/parser/Parser.java]: I think checking the condition this way causes inputs such as "todo bookmark page 50" to enter the wrong branch

File: src/main/java/usagi/parser/Parser.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+                    LocalDateTime from = Task.parseDateTimeFlexible(p2[0].trim());
+                    LocalDateTime to   = Task.parseDateTimeFlexible(p2[1].trim());
+
+                    Task task = new Event(p1[1], false, from, to);
+                    tasks.add(task);
+                    storage.save(tasks.all());
+                    ui.showAdded(task, tasks.size());
+                } else {
+                    ui.showLine("Ura? (deadline must be followed by a description and a /from and /to)");
```
Comment (jooh6969) [inline src/main/java/usagi/parser/Parser.java]: Should this be event instead of deadline?

File: src/main/java/usagi/storage/Storage.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+     * Ensures that the parent directories of the given path exist.
+     * Creates them if they do not already exist.
+     *
+     * @param p The file path whose parent directories should be checked
+     * @return The same path p
+     * @throws IOException If the directories cannot be created
+     */
+    private static Path ensureParentDirs(Path p) throws IOException {
+        Path parent = p.getParent();
+        if (parent != null) Files.createDirectories(parent);
```
Comment (jooh6969) [inline src/main/java/usagi/storage/Storage.java:49]: I think this if block should follow egyptian style parenthesis even though it's only one line

File: src/main/java/usagi/storage/Storage.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+     * Saves the given list of tasks to the storage file.
+     * 
+     * Each task is serialized into a line using Task.toLine().
+     * The file is created if it doesn't exist, or overwritten if it does.
+     *
+     * @param tasks The list of tasks to save
+     * @throws UsagiException If an error occurs during file writing
+     */
+    public void save(List<Task> tasks) throws UsagiException {
+        try {
```
Comment (jooh6969) [inline src/main/java/usagi/storage/Storage.java:90]: Excellent indentation, makes the code clean and easily readable

File: src/main/java/usagi/task/ToDos.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
+package usagi.task;
+
+/**
+ * Represents a Todo task with title and completion status.
+ * This is a concrete class inherited from the abstract Task class.
+ */
+public class ToDos extends Task {
```
Comment (NownuJ) [inline src/main/java/usagi/task/ToDos.java:7]: Todo would be a better name for this class. ToDos suggests that this class is like a container for multiple todo tasks

File: src/main/java/usagi/task/Deadline.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+ * Represents a Deadline task with title, completion status and a by time.
+ * This is a concrete class inherited from the abstract Task class.
+ */
+public class Deadline extends Task {
+    protected LocalDateTime by;
+
+    /**
+     * DateTimeFormatter for formatting LocalDateTime objects for UI display.
+     */
+    private static final DateTimeFormatter UI = DateTimeFormatter.ofPattern("MMM d yyyy HH:mm");
```
Comment (NownuJ) [inline src/main/java/usagi/task/Deadline.java:17]: Great name for UI variable that clearly shows that it helps to format DateTime!

File: src/main/java/usagi/task/Deadline.java [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+        this.by = by;
+    }
+
+    @Override
+    public String type() {
+        return "D";
+    }
+
+    @Override
+    public String[] extra() {
```
Comment (NownuJ) [inline src/main/java/usagi/task/Deadline.java:38]: For arrays, they should be named plural like 'extras'. Same goes for all other tasks with 'extra' method

File: src/main/java/usagi/task/Deadline.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+     * @param done The completion status of the task
+     * @param by The due date and time for the task
+     */
+    public Deadline(String title, boolean done, LocalDateTime by) {
+        super(title, done);
+        this.by = by;
+    }
+
+    @Override
+    public String type() {
```
Comment (NownuJ) [inline src/main/java/usagi/task/Deadline.java:33]: Methods should be named like an action, such as getType()

File: src/main/java/usagi/task/TaskList.java [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+    public TaskList(List<Task> tasks) {
+        this.tasks = new ArrayList<>(tasks);
+    }
+
+    /**
+     * Returns all tasks in the list.
+     * 
+     * @return A list containing all tasks
+     */
+    public List<Task> all() {
```
Comment (NownuJ) [inline src/main/java/usagi/task/TaskList.java:40]: Once again, methods should be named like an action, getAll() in this case.

There are a couple of methods that are named after what they return, I think it would be a quick fix to add 'get-' prefix on them.

File: src/main/java/usagi/ui/Ui.java [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+    }
+
+    /**
+     * Displays an error message when the user input cannot be understood.
+     */
+    public void showUnknown() {
+        System.out.println("Ura? (Usagi cannot understand ur gibberish)");
+    }
+}
+
```
Comment (NownuJ) [inline src/main/java/usagi/ui/Ui.java]: Great namings for all the methods here!

**Review by NownuJ — COMMENTED** [review]
Good code quality, I think some method names can be changed to be like an action.
